integration/ksm/ksm_test.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+#
+# Copyright (c) 2019 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+set -e
+
+dir_path=$(dirname "$0")
+source "${dir_path}/../../lib/common.bash"
+source "${dir_path}/../../metrics/lib/common.bash"
+
+TEST_NAME="Run containers with KSM"
+KATA_KSM_THROTTLER="${KATA_KSM_THROTTLER:-yes}"
+RUNTIME="${RUNTIME:-kata-runtime}"
+PAYLOAD_ARGS="${PAYLOAD_ARGS:-tail -f /dev/null}"
+IMAGE="${IMAGE:-busybox}"
+WAIT_TIME="60"
+
+function turn_on_ksm() {
+	echo "Turning on KSM"
+	echo "1" | sudo tee "${KSM_ENABLE_FILE}"
+}
+
+function setup() {
+	clean_env
+
+	check_processes
+
+	turn_on_ksm
+
+	check_for_ksm
+}
+
+function teardown() {
+	clean_env
+
+	check_processes
+
+	save_ksm_settings
+
+	restore_ksm_settings
+
+	disable_ksm
+}
+
+function run_with_ksm() {
+	setup
+
+	# Running the first container
+	docker run -d --runtime="${RUNTIME}" "${IMAGE}" sh -c "${PAYLOAD_ARGS}"
+
+	echo "Entering KSM settle mode on first container"
+	wait_ksm_settle "${WAIT_TIME}"
+
+	# Checking the pages merged for the first container
+	first_pages_merged=$(cat "${KSM_PAGES_SHARED}")
+
+	echo "Pages merged $first_pages_merged"
+
+	# If KSM is enabled, then you normally want to sleep long enough to
+	# let it do its work and for the numbers to 'settle'.
+	echo "napping ${WAIT_TIME} s"
+	sleep "${WAIT_TIME}"
+
+	# Running the second container
+	docker run -d --runtime="${RUNTIME}" "${IMAGE}" sh -c "${PAYLOAD_ARGS}"
+
+	echo "Entering KSM settle mode on second container"
+	wait_ksm_settle "${WAIT_TIME}"
+
+	# Checking the pages merged for the second container
+	second_pages_merged=$(cat "${KSM_PAGES_SHARED}")
+
+	echo "Pages merged $second_pages_merged"
+
+	echo "napping ${WAIT_TIME} s"
+	sleep "${WAIT_TIME}"
+
+	# Compared the pages merged between the containers
+	echo "Comparing merged pages between containers"
+	[ "$second_pages_merged" -gt "$first_pages_merged" ] || die "The merged pages on the second container is less than the first container"
+
+	teardown
```

Comment:
I think we should use `trap` to make sure the `teardown` runs if failures occur before

Reply:
done @chavafg